Call OnNativeWidgetDestroying for native_widget_delegate_ on closing

jkim-julie opened the pull request:
This patch makes content_browsertests work. Because
DesktopWindowTreeHostPlatform missed calling OnNativeWidgetDestroying()
for |native_widget_delegate_|, it couldn't destroy shell properly.
With this patch, we can run content_browsertests with headless.

Issue #325

jkim-julie commented:
@msisov, we can run content_browsertests with headless with this patch and we have only one failed test, OutOfProcessPPAPITest.TrueTypeFont (../../content/test/ppapi/ppapi_browsertest.cc:141). PTAL.

jkim-julie commented:
I found that WaylandWindow calls OnAcceleratedWidgetDestroying() for itself on dtor, but other platform windows don't. Even if other DesktopWindowTreeHostXXX calls 'native_widget_delegate_->OnNativeWidgetDestroying()'  on CloseNow(), Ozone could change the sequence like WaylandWindow did. If then, we need to add calling OnAcceleratedWidgetDestroying() at each platform window. WDYT?

jkim-julie commented:
After I looked into ScenicWindow, I surely realized that we need to add 'OnAcceleratedWidgetDestroying/Destroyed' to each platform window on destructing. I updated the patch with calling OnAcceleratedWidgetDestroying/Destroyed. With this, we can run content_browsertests on headless/x11/wayland. PTAL.

msisov commented:
> I found that WaylandWindow calls OnAcceleratedWidgetDestroying() for itself on dtor, but other platform windows don't. Even if other DesktopWindowTreeHostXXX calls 'native_widget_delegate_->OnNativeWidgetDestroying()' on CloseNow(), Ozone could change the sequence like WaylandWindow did. If then, we need to add calling OnAcceleratedWidgetDestroying() at each platform window. WDYT?

You're right. After I've modified the desktop_window_tree_host_platform, it became a requirement. What is more, when doing like that, we ensure a correct destruction order.

> After I looked into ScenicWindow, I surely realized that we need to add 'OnAcceleratedWidgetDestroying/Destroyed' to each platform window on destructing. I updated the patch with calling OnAcceleratedWidgetDestroying/Destroyed. With this, we can run content_browsertests on headless/x11/wayland. PTAL.

Awesome!

msisov commented on ui/ozone/platform/drm/host/drm_window_host.cc:
```diff
@@ -41,12 +41,14 @@ DrmWindowHost::DrmWindowHost(PlatformWindowDelegate* delegate,
 }
 
 DrmWindowHost::~DrmWindowHost() {
+  delegate_->OnAcceleratedWidgetDestroying();
```
not needed here, because Drm doesn't take the desktop_window_tree_host_platform path.

msisov commented on ui/ozone/platform/cast/platform_window_cast.cc:
```diff
@@ -25,8 +25,10 @@ PlatformWindowCast::PlatformWindowCast(PlatformWindowDelegate* delegate,
 }
 
 PlatformWindowCast::~PlatformWindowCast() {
+  delegate()->OnAcceleratedWidgetDestroying();
```
not sure about this. better avoid touching this cast platform.

msisov commented on ui/platform_window/win/win_window.cc:
```diff
@@ -56,6 +56,8 @@ WinWindow::WinWindow(PlatformWindowDelegate* delegate, const gfx::Rect& bounds)
 }
 
 WinWindow::~WinWindow() {
```
not sure about this. I'd rather kept it without this.

msisov commented on ui/platform_window/x11/x11_window_base.cc:
```diff
@@ -97,7 +97,9 @@ X11WindowBase::X11WindowBase(PlatformWindowDelegate* delegate,
 X11WindowBase::~X11WindowBase() {
+  delegate_->OnAcceleratedWidgetDestroying();
   UnConfineCursor();
+  delegate_->OnAcceleratedWidgetDestroyed();
```
place it after destroy? Wdyt? Can you check what's the best destruction order here?

msisov commented on ui/views/widget/desktop_aura/desktop_window_tree_host_platform.cc:
```diff
@@ -139,6 +139,14 @@ void DesktopWindowTreeHostPlatform::Close() {
+  // It checks IsVisible() here in case that it's called directly and if it's
+  // visible, Hide() is called.
+  if (compositor()->IsVisible())
+    Hide();
```
Can you share with me the cases when CloseNow() can be called without ::Close() call?

jkim-julie replied:
Thanks for the review. Here is the case, which is the callstack when we run content_browsertests on wayland now.
[109443:109443:0621/180918.879416:FATAL:compositor.cc(499)] Check failed: !IsVisible(). 
#0 0x7fba8626a3bc base::debug::StackTrace::StackTrace()
#1 0x7fba861b465b logging::LogMessage::~LogMessage()
#2 0x7fba8695541e ui::Compositor::ReleaseAcceleratedWidget()
#3 0x7fba86440c5a aura::WindowTreeHostPlatform::OnAcceleratedWidgetDestroyed()
#4 0x7fba7ba13866 ui::HeadlessWindow::~HeadlessWindow()
#5 0x7fba7c752a30 views::DesktopWindowTreeHostPlatform::CloseNow()
#6 0x7fba7c730472 views::Widget::CloseNow()
#7 0x000000daf14a content::Shell::CloseAllWindows()
#8 0x000000d0ea2c content::ContentBrowserTest::PostRunTestOnMainThread()
#9 0x000000d19cee content::BrowserTestBase::ProxyRunTestOnMainThreadLoop()
#10 0x000000db22d6 content::ShellBrowserMainParts::PreMainMessageLoopRun()
#11 0x7fba83f1e051 content::BrowserMainLoop::PreMainMessageLoopRun()
#12 0x7fba84415125 content::StartupTaskRunner::RunAllTasksNow()
#13 0x7fba83f1c74c content::BrowserMainLoop::CreateStartupTasks()
#14 0x7fba83f20ae5 content::BrowserMainRunnerImpl::Initialize()
#15 0x7fba83f20a47 content::BrowserMainRunnerImpl::Initialize()
#16 0x000000db1add ShellBrowserMain()
#17 0x000000d6763d content::ShellMainDelegate::RunProcess()
#18 0x7fba849efe43 content::RunBrowserProcessMain()
#19 0x7fba849f0f96 content::ContentMainRunnerImpl::Run()
#20 0x7fba81ad96bf service_manager::Main()
#21 0x7fba849eed34 content::ContentMain()
#22 0x000000d19662 content::BrowserTestBase::SetUp()
#23 0x000000d0e7f4 content::ContentBrowserTest::SetUp()
#24 0x000000c06e1d testing::Test::Run()
#25 0x000000c07a80 testing::TestInfo::Run()
#26 0x000000c07f97 testing::TestCase::Run()
#27 0x000000c13987 testing::internal::UnitTestImpl::RunAllTests()
#28 0x000000c134fd testing::UnitTest::Run()
#29 0x000000d54da1 base::TestSuite::Run()
#30 0x000000d12587 content::ContentTestLauncherDelegate::RunTestSuite()
#31 0x000000d3a6d0 content::LaunchTests()

msisov replied:
What if you just move the call to Hide() from the DesktopWindowTreeHostPlatform::Close() to 
DesktopWindowTreeHostPlatform::CloseNow() ?

For example,

```
DesktopWindowTreeHostPlatform::CloseNow() {
.....
 if (compositor()->IsVisible())
   Hide();
                                              
  SetPlatformWindow(nullptr);                                                   
  if (!weak_ref || got_on_closed_)                                              
    return;    
.....
}
```

jkim-julie committed:
Add OnAcceleratedWidgetDestroying/Destroyed on platform windows

This patch makes content_browsertests work. Platform windows
should call OnAcceleratedWidgetDestroying/Destroyed on
destructing to delete resources for windows.

It also moves Hide() to DesktopWindowTreeHostPlatform::CloseNow()
from DesktopWindowTreeHostPlatform::Close() because CloseNow()
could be called directly and removes delegate_->OnClosed() in
X11WindowBase because it's handled in
DesktopWindowTreeHostPlatform::CloseNow().

With this patch, we can run content_browsertests with headless/X11/
Wayland.

Issue #325

jkim-julie commented:
@msisov, I uploaded the updated patch.
It keeps OnAcceleratedWidgetXXX calls only in HeadlessWindow and X11WindowBase, moves Hide() to CloseNow() and removes 'delegate_->OnClose()' in X11WindowBase.

I checked the destruction sequence and it shows that DesktopWindowTreeHostPlatform::CloseNow() is called and it resets PlatformWindow and calls desktop_native_widget_aura_->OnHostClosed() for itself. So, I think platform window doesn't need to call 'delegate_->OnClosed'.

* delegate_->OnClosed() couldn't be called before 'delegate_->OnAcceleratedWidgetDestroyed()' because OnClose() calls 'desktop_native_widget_aura_->OnHostClosed()' and it destroys DesktopWindowTreeHostPlatform.

PTAL.

msisov commented:
lgtm
